File: DevExpress VCL/ExpressPrinting System/Demos/CBuilder/EQGrid4RL/EQGrid4RLMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#include "shellapi.hpp"
#pragma hdrstop

#include "EQGrid4RLMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBTableView"
#pragma link "cxGridLevel"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxEditRepositoryItems"
#pragma link "dxPSCore"
#pragma link "dxPScxGrid6Lnk"
#pragma link "dxBkgnd"
#pragma link "dxPrnDev"
#pragma link "dxPrnPg"
#pragma link "dxPSCompsProvider"
#pragma link "dxPSEdgePatterns"
#pragma link "dxPSEngn"
#pragma link "dxPSFillPatterns"
#pragma link "dxPSGlbl"
#pragma link "dxPSUtl"
#pragma link "dxWrap"
#pragma link "cxGridCustomView"
#pragma resource "*.dfm"
TEQGrid4RLMainForm *EQGrid4RLMainForm;
//---------------------------------------------------------------------------
__fastcall TEQGrid4RLMainForm::TEQGrid4RLMainForm(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actLookAndFeelKindUltraFlatExecute(TObject *Sender)
{
  if(cxGrid->LookAndFeel->Kind != lfUltraFlat){
    cxGrid->LookAndFeel->Kind = lfUltraFlat;
    ((TMenuItem*)Sender)->Checked = true;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actLookAndFeelKindFlatExecute(TObject *Sender)
{
  if(cxGrid->LookAndFeel->Kind != lfFlat){
    cxGrid->LookAndFeel->Kind = lfFlat;
    ((TMenuItem*)Sender)->Checked = true;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actLookAndFeelKindStandardExecute(TObject *Sender)
{
  if(cxGrid->LookAndFeel->Kind != lfStandard){
    cxGrid->LookAndFeel->Kind = lfStandard;
    ((TMenuItem*)Sender)->Checked = true;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::miNativeStyleClick(TObject *Sender)
{
  cxGrid->LookAndFeel->NativeStyle = !cxGrid->LookAndFeel->NativeStyle;
  ((TCustomAction*)Sender)->Checked = !((TCustomAction*)Sender)->Checked;
}
//---------------------------------------------------------------------------



void __fastcall TEQGrid4RLMainForm::actShowDemoDescriptionExecute(
      TObject *Sender)
{
  lbDescrip->Visible = !lbDescrip->Visible;
  ((TCustomAction*)Sender)->Checked = !((TCustomAction*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actGridNativeStyleExecute(
      TObject *Sender)
{
  cxGrid->LookAndFeel->NativeStyle = !cxGrid->LookAndFeel->NativeStyle;
  ((TCustomAction*)Sender)->Checked = !((TCustomAction*)Sender)->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actDownloadsExecute(TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://www.devexpress.com/downloads/index.asp", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actForumExecute(TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://24.234.251.34/dxforum/dxforumisapi.dll/", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actDXOnTheWebExecute(
      TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://www.devexpress.com/index.shtm", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actProductsExecute(TObject *Sender)
{
  ShellExecute(Handle, "OPEN", "http://www.devexpress.com/products/index.asp", NULL, NULL, SW_SHOWMAXIMIZED);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::AlwaysEnabled(TObject *Sender)
{
  ((TCustomAction*)Sender)->Enabled = true;
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actExitExecute(TObject *Sender)
{
  Close();  
}
//---------------------------------------------------------------------------


void __fastcall TEQGrid4RLMainForm::CustomizeColumns()
{
  const cDistance = 3;
  const cPeriod = 4;
  const cRadius = 7;
  DecimalSeparator = '.';
  for(int i=0; i < tvPlanets->ColumnCount; i++)
    if((i == cDistance) || (i == cRadius))
      tvPlanets->Columns[i]->DataBinding->ValueTypeClass = __classid(TcxIntegerValueType);
    else
      if (i == cPeriod)
        tvPlanets->Columns[i]->DataBinding->ValueTypeClass = __classid(TcxFloatValueType);
      else
        tvPlanets->Columns[i]->DataBinding->ValueTypeClass = __classid(TcxStringValueType);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::InitRecord(String const Str, int AInt, TStringList* AValues)
{
  Variant V;
  AValues->CommaText = Str;
  for(int i=0; i < AValues->Count; i++)
   if (AValues->Strings[i] != "-") {
     V = AValues->Strings[i];
      if (!VarIsNull(V)) {
        if (tvPlanets->Columns[i]->DataBinding->ValueTypeClass == __classid(TcxFloatValueType))
          V = VarAsType(StrToFloat(V), varDouble);
        tvPlanets->DataController->Values[AInt][i] = V;
      }
   }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::LoadData()
{
  const String AFileName = "nineplanets.txt";
  const int AHeaderLineCount = 2;

  if (!FileExists(AFileName))
    throw Exception("Data file not found");

  TStringList* ARecords = new TStringList();
  TStringList* AValues = new TStringList();
  try {
    ARecords->LoadFromFile(AFileName);
    tvPlanets->BeginUpdate();
    tvPlanets->DataController->RecordCount = ARecords->Count - AHeaderLineCount;
    for (int i=0; i < ARecords->Count - (AHeaderLineCount + 1) + 1; i++)
      InitRecord(ARecords->Strings[i + AHeaderLineCount], i, AValues);
  }
  __finally {
    tvPlanets->EndUpdate();
    delete ARecords;
    delete AValues;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::FormCreate(TObject *Sender)
{
  CustomizeColumns();
  LoadData();
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::FormShow(TObject *Sender)
{
  tvPlanets->DataController->Groups->FullExpand();
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actPageSetupExecute(TObject *Sender)
{
  dxComponentPrinter->PageSetup(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actPreviewExecute(TObject *Sender)
{
  dxComponentPrinter->Preview(true,NULL);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actPrintExecute(TObject *Sender)
{
  dxComponentPrinter->Print(true,NULL,NULL);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actDesignerExecute(TObject *Sender)
{
  dxComponentPrinter->DesignReport(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actFullExpandExecute(TObject *Sender)
{
  tvPlanets->DataController->Groups->FullExpand();
}
//---------------------------------------------------------------------------

void __fastcall TEQGrid4RLMainForm::actFullCollapseExecute(TObject *Sender)
{
  tvPlanets->DataController->Groups->FullCollapse();
}
//---------------------------------------------------------------------------

 

File: DevExpress VCL/ExpressPrinting System/Demos/Delphi/EQGrid4RL/EQGrid4RLMain.pas
unit EQGrid4RLMain;

{$I ..\cxVer.inc}

interface

uses
  Windows, Messages, SysUtils{$IFDEF DELPHI6}, Variants{$ENDIF}, Classes, Controls,
  Forms, cxStyles, cxCustomData, cxGraphics, cxFilter, cxData, cxEdit,
  cxEditRepositoryItems, cxGridTableView,  ActnList, ImgList,
  Menus, cxGridLevel, cxGridCustomTableView, cxClasses, Dialogs,
  cxControls, cxGridCustomView, cxGrid, ComCtrls, StdCtrls, ShellAPI,
  dxPSGlbl, dxPSUtl, dxPSEngn, dxPrnPg, dxBkgnd, dxWrap, dxPrnDev,
  dxPSCompsProvider, dxPSFillPatterns, dxPSEdgePatterns, dxPSCore,
  dxPScxGrid6Lnk, cxDataStorage, ToolWin;

type
  TEQGrid4RLMainForm = class(TForm)
    mmMain: TMainMenu;
    miHelp: TMenuItem;
    miDeveloperExpressontheweb: TMenuItem;
    miForum: TMenuItem;
    miDownloads: TMenuItem;
    miProducts: TMenuItem;
    ilMain: TImageList;
    sty: TActionList;
    actProducts: TAction;
    actDownloads: TAction;
    actForum: TAction;
    actDXOnTheWeb: TAction;
    miFile: TMenuItem;
    miOptions: TMenuItem;
    miExit: TMenuItem;
    actExit: TAction;
    sbMain: TStatusBar;
    actShowDemoDescription: TAction;
    miShowDemoDescription: TMenuItem;
    actGridNativeStyle: TAction;
    cxGrid: TcxGrid;
    miGridLookFeel: TMenuItem;
    miKind: TMenuItem;
    miFlat: TMenuItem;
    miStandard: TMenuItem;
    miUltraFlat: TMenuItem;
    miNativeStyle: TMenuItem;
    miSeparator2: TMenuItem;
    StyleRepository: TcxStyleRepository;
    tvssDevExpress: TcxGridTableViewStyleSheet;
    cxStyle1: TcxStyle;
    cxStyle2: TcxStyle;
    cxStyle3: TcxStyle;
    cxStyle4: TcxStyle;
    cxStyle5: TcxStyle;
    cxStyle6: TcxStyle;
    cxStyle7: TcxStyle;
    cxStyle8: TcxStyle;
    cxStyle9: TcxStyle;
    cxStyle10: TcxStyle;
    cxStyle11: TcxStyle;
    cxStyle12: TcxStyle;
    cxStyle13: TcxStyle;
    cxStyle14: TcxStyle;
    lvPlanets: TcxGridLevel;
    tvPlanets: TcxGridTableView;
    lbDescrip: TLabel;
    tvPlanetsNAME: TcxGridColumn;
    tvPlanetsNO: TcxGridColumn;
    tvPlanetsORBITS: TcxGridColumn;
    tvPlanetsDISTANCE: TcxGridColumn;
    tvPlanetsPERIOD: TcxGridColumn;
    tvPlanetsDISCOVERER: TcxGridColumn;
    tvPlanetsDATE: TcxGridColumn;
    tvPlanetsRADIUS: TcxGridColumn;
    edrepMain: TcxEditRepository;
    edrepCenterText: TcxEditRepositoryTextItem;
    edrepRightText: TcxEditRepositoryTextItem;
    dxComponentPrinter: TdxComponentPrinter;
    dxComponentPrinterLink: TdxGridReportLink;
    ToolBar1: TToolBar;
    tbtnPageSetup: TToolButton;
    tbtnPreview: TToolButton;
    tbtnPrint: TToolButton;
    tbtnDesigner: TToolButton;
    N1: TMenuItem;
    miPirntingOptions: TMenuItem;
    N2: TMenuItem;
    Print1: TMenuItem;
    miPrintPreview: TMenuItem;
    miPageSetup: TMenuItem;
    actPageSetup: TAction;
    actPreview: TAction;
    actPrint: TAction;
    actDesigner: TAction;
    N3: TMenuItem;
    miFullExpand: TMenuItem;
    miFullCollapsing: TMenuItem;
    actFullExpand: TAction;
    actFullCollapse: TAction;
    ToolButton1: TToolButton;
    tbtnFullCollapse: TToolButton;
    tbtnFullExpand: TToolButton;
    procedure AlwaysEnabled(Sender: TObject);
    procedure actProductsExecute(Sender: TObject);
    procedure actDownloadsExecute(Sender: TObject);
    procedure actForumExecute(Sender: TObject);
    procedure actDXOnTheWebExecute(Sender: TObject);
    procedure actExitExecute(Sender: TObject);
    procedure actShowDemoDescriptionExecute(Sender: TObject);
    procedure actGridNativeStyleExecute(Sender: TObject);
    procedure FormShow(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure miUltraFlatClick(Sender: TObject);
    procedure miStandardClick(Sender: TObject);
    procedure miFlatClick(Sender: TObject);
    procedure actPageSetupExecute(Sender: TObject);
    procedure actPreviewExecute(Sender: TObject);
    procedure actPrintExecute(Sender: TObject);
    procedure actDesignerExecute(Sender: TObject);
    procedure actFullExpandExecute(Sender: TObject);
    procedure actFullCollapseExecute(Sender: TObject);
  private
    procedure CustomizeColumns;
    procedure LoadData;
  end;

var
  EQGrid4RLMainForm: TEQGrid4RLMainForm;

implementation

uses
  cxLookAndFeels;

{$R *.dfm}

procedure TEQGrid4RLMainForm.AlwaysEnabled(Sender: TObject);
begin
  TCustomAction(Sender).Enabled := True;
end;

procedure TEQGrid4RLMainForm.actProductsExecute(Sender: TObject);
begin
  ShellExecute(Handle, PChar('OPEN'), PChar('http://www.devexpress.com/products/index.asp'), nil, nil, SW_SHOWMAXIMIZED);
end;

procedure TEQGrid4RLMainForm.actDownloadsExecute(Sender: TObject);
begin
  ShellExecute(Handle, PChar('OPEN'), PChar('http://www.devexpress.com/downloads/index.asp'), nil, nil, SW_SHOWMAXIMIZED);
end;

procedure TEQGrid4RLMainForm.actForumExecute(Sender: TObject);
begin
  ShellExecute(Handle, PChar('OPEN'), PChar('http://24.234.251.34/dxforum/dxforumisapi.dll/'), nil, nil, SW_SHOWMAXIMIZED);
end;

procedure TEQGrid4RLMainForm.actDXOnTheWebExecute(
  Sender: TObject);
begin
  ShellExecute(Handle, PChar('OPEN'), PChar('http://www.devexpress.com/index.shtm'), nil, nil, SW_SHOWMAXIMIZED);
end;

procedure TEQGrid4RLMainForm.actExitExecute(Sender: TObject);
begin
  Close;
end;

procedure TEQGrid4RLMainForm.actShowDemoDescriptionExecute(
  Sender: TObject);
begin
  lbDescrip.Visible := not lbDescrip.Visible;
  TCustomAction(Sender).Checked := not TCustomAction(Sender).Checked;
end;

procedure TEQGrid4RLMainForm.actGridNativeStyleExecute(Sender: TObject);
begin
  cxGrid.LookAndFeel.NativeStyle := not cxGrid.LookAndFeel.NativeStyle;
  TCustomAction(Sender).Checked := cxGrid.LookAndFeel.NativeStyle; 
end;

procedure TEQGrid4RLMainForm.FormShow(Sender: TObject);
begin
  tvPlanets.DataController.Groups.FullExpand;
end;


procedure TEQGrid4RLMainForm.FormCreate(Sender: TObject);
begin
  CustomizeColumns;
  LoadData;
end;


procedure TEQGrid4RLMainForm.CustomizeColumns;
const
  cDistance = 3;
  cPeriod = 4;
  cRadius = 7;
var
  I: Integer;
begin
  DecimalSeparator := '.';
  with tvPlanets do
  for I := 0 to ColumnCount - 1 do
    if I in [cDistance, cRadius] then
      Columns[I].DataBinding.ValueTypeClass := TcxIntegerValueType
    else
      if I in [cPeriod] then
      Columns[I].DataBinding.ValueTypeClass := TcxFloatValueType
      else
       Columns[I].DataBinding.ValueTypeClass := TcxStringValueType;
end;

procedure TEQGrid4RLMainForm.LoadData;
const
  AFileName = 'nineplanets.txt';
  AHeaderLineCount = 2;

var
  ARecords, AValues: TStringList;
  I: Integer;

  procedure InitRecord(const Str: string);
  var
    J: Integer;
    V: Variant;
  begin
    AValues.CommaText := Str;
    for J := 0 to AValues.Count - 1 do
     if AValues.Strings[J] <> '-' then
     begin
      V := AValues.Strings[J];
      if not VarIsNull(V) then
      begin
        if tvPlanets.Columns[J].DataBinding.ValueTypeClass = TcxFloatValueType then
          V := StrToFloat(V);
        tvPlanets.DataController.Values[I, J] := V;
      end;
     end;
  end;

begin
  if not FileExists(AFileName) then
    raise Exception.Create('Data file not found');

  ARecords := TStringList.Create;
  AValues := TStringList.Create;

  with ARecords do
  try
    LoadFromFile(AFileName);
    tvPlanets.BeginUpdate;
    tvPlanets.DataController.RecordCount := Count - AHeaderLineCount;
    for I := 0 to Count - (AHeaderLineCount + 1) do
      InitRecord(Strings[I + AHeaderLineCount]);
  finally
    tvPlanets.EndUpdate;
    ARecords.Free;
    AValues.Free;
  end;
end;

procedure TEQGrid4RLMainForm.miUltraFlatClick(Sender: TObject);
begin
  cxGrid.LookAndFeel.Kind := lfUltraFlat;
  TMenuItem(Sender).Checked := True;
end;

procedure TEQGrid4RLMainForm.miStandardClick(Sender: TObject);
begin
  cxGrid.LookAndFeel.Kind := lfStandard;
  TMenuItem(Sender).Checked := True;
end;

procedure TEQGrid4RLMainForm.miFlatClick(Sender: TObject);
begin
  cxGrid.LookAndFeel.Kind := lfFlat;
  TMenuItem(Sender).Checked := True;
end;

procedure TEQGrid4RLMainForm.actPageSetupExecute(Sender: TObject);
begin
  dxComponentPrinter.PageSetup(nil);
end;

procedure TEQGrid4RLMainForm.actPreviewExecute(Sender: TObject);
begin
  dxComponentPrinter.Preview(True,nil);
end;

procedure TEQGrid4RLMainForm.actPrintExecute(Sender: TObject);
begin
  dxComponentPrinter.Print(True,nil,nil);
end;

procedure TEQGrid4RLMainForm.actDesignerExecute(Sender: TObject);
begin
  dxComponentPrinter.DesignReport(nil);
end;

procedure TEQGrid4RLMainForm.actFullExpandExecute(Sender: TObject);
begin
  tvPlanets.DataController.Groups.FullExpand;
end;

procedure TEQGrid4RLMainForm.actFullCollapseExecute(Sender: TObject);
begin
  tvPlanets.DataController.Groups.FullCollapse;
end;

end.
